Jack/horizontal cv conversion

jschluger opened the pull request:
Addressing issue #145.

kevinchan159 commented on Clicker/SectionControllers/EmptyStateSectionController.swift:
```diff
@@ -23,6 +28,10 @@ class EmptyStateSectionController: ListSectionController {
         let cell = collectionContext?.dequeueReusableCell(of: EmptyStateCell.self, for: self, at: index) as! EmptyStateCell
+        cell.session = session
+        cell.userRole = userRole
+        cell.nameViewDelegate = nameViewDelegate
```
instead of assigning them directly here, we should either
1) make an `EmptyStateCellDelegate` protocol which has `session`, `userRole`, and `nameViewDelegate`
2) or have a `configure(with: session: Session, userRole: UserRole, nameViewDelegate: NameViewDelegate)` function in EmptyStateCell

jschluger replied:
I will clean up the places where I did this.

kevinchan159 commented on Clicker/SectionControllers/PollSectionController.swift:
```diff
@@ -9,15 +9,24 @@
+    var session: Session!
+    var userRole: UserRole!
+    var socket: Socket!
+    var askedCardDelegate: AskedCardDelegate!
```
same question as above, where are these set?

kevinchan159 commented on Clicker/SectionControllers/PollSectionController.swift:
```diff
@@ -9,15 +9,24 @@
     
+    override init() {
+        super.init()
+        self.inset = UIEdgeInsets(top: 0, left: 0, bottom: 0, right: 10)
```
what is this `inset` being used for?

jschluger replied:
The inset between the cards. Without it the cards appeared directly in contact with each other.

kevinchan159 commented on Clicker/SectionControllers/PollSectionController.swift:
```diff
@@ -27,12 +36,30 @@ class PollSectionController: ListSectionController {
+            socket.addDelegate(cell)
+            cell.poll = poll
+            cell.cardType = poll.state
+            cell.configure()
```
same as comment above, either make a `AskedCardDelegate` or make a `configure(with ...)` function which contains these params

kevinchan159 commented on Clicker/SectionControllers/PollSectionController.swift:
```diff
@@ -27,12 +36,30 @@ class PollSectionController: ListSectionController {
+            socket.addDelegate(cell)
+            cell.poll = poll
+            cell.cardType = poll.state
+            cell.configure()
```
same^

kevinchan159 commented on Clicker/ViewControllers/CardController+Extension.swift:
```diff
@@ -31,11 +34,22 @@ extension CardController: ListAdapterDataSource {
+            pollSectionController.session = session
+            pollSectionController.userRole = userRole
+            pollSectionController.socket = socket
+            pollSectionController.askedCardDelegate = self
```
pass these variables through the init of PollSectionController

AAAstorga replied:
+1

kevinchan159 commented on Clicker/ViewControllers/CardController+Extension.swift:
```diff
@@ -31,11 +34,22 @@ extension CardController: ListAdapterDataSource {
+            let emptyStateController = EmptyStateSectionController()
+            emptyStateController.session = session
+            emptyStateController.userRole = userRole
+            emptyStateController.nameViewDelegate = self
```
pass these variables through the init of EmptyStateSectionController

kevinchan159 commented on Clicker/Views/Cards/AnswerCard.swift:
```diff
@@ -18,7 +18,7 @@ class AnswerCard: UICollectionViewCell, CardDelegate, SocketDelegate {
     var poll: Poll!
     var socket: Socket!
-    var cardType: CardType!
+    var cardType: PollState!
```
do we need this if we have the `poll` variable which has `PollState`?

jschluger replied:
You're right, that would be a smarter implementation. We previously had a `CardType` enum with the same exact values as the `PollState` enum so i removed the `CardType` enum and replaced it with `PollState` where it was previously being used, but didn't realize we are already storing this information in the `poll` object.

kevinchan159 replied:
Oh yeah I remember we had that when we were doing stuff for BOOM haha

kevinchan159 commented on Clicker/Views/Cards/EmptyStateCell.swift:
```diff
@@ -18,4 +42,71 @@ class EmptyStateCell: UICollectionViewCell {
+    func setup() {
+        setupViews()
+        setupConstraints()
+        print("setting up empty state cell with userRole = \(userRole), session=\(session)")
```
remove print statement

kevinchan159 commented on Clicker/Views/Cards/EmptyStateCell.swift:
```diff
@@ -18,4 +42,71 @@ class EmptyStateCell: UICollectionViewCell {
+        waitingLabel.lineBreakMode = .byWordWrapping
+        waitingLabel.numberOfLines = 0
+        waitingLabel.text = userRole == .admin ? adminWaitingText : userWaitingText
+        addSubview(waitingLabel)
```
lets add these as subviews to `contentView` instead

jschluger replied:
Sure.

kevinchan159 commented on Clicker/Views/Cards/EmptyStateCell.swift:
```diff
@@ -18,4 +42,71 @@ class EmptyStateCell: UICollectionViewCell {
+        nameView.session = session
+        nameView.delegate = nameViewDelegate
+        nameView.setup()
+        addSubview(nameView)
```
add as subview to `contentView`

kevinchan159 commented on Clicker/Views/Cards/EmptyStateCell.swift:
```diff
@@ -18,4 +42,71 @@ class EmptyStateCell: UICollectionViewCell {
+    func setupNameView() {
+        nameView = NameView(frame: .zero)
+        nameView.session = session
+        nameView.delegate = nameViewDelegate
```
can we make a custom init for `NameView` where we pass these args in or have a `configure(with ..)` function?

jschluger commented on Clicker/Models/ListDiffableModels/PollDateModel.swift:
```diff
@@ -13,11 +13,13 @@ class PollDateModel {
+    init(date: String, poll: Poll, index: Int) {
         self.date = date
         self.poll = poll
+        self.index = index
```
This is used in `didSelectItemAt` so that when you press on a cell in the vertical view, the polls for the correct date are opened in the horizontal view: `currentIndex` for the data source in the horizontal view is set to this `index`.

kevinchan159 replied:
I don't think having the `index` as a part of the model is that great because the index can change. Instead, inside the section controller's `didSelectItem` function, we could pass back the `PollDateModel` associated with it back to `CardController` to find the `index`. You don't have to add this in this PR though, just letting you know how we would do it.

jschluger commented:
In the latest commit I addressed feedback on this PR: 
 - Added `configureWith(...)` to places where I had previously been setting lots of variables in new instances of classes manually. (As a side note, this might be worth doing across the project as part of our refactoring).
 - Completely removed `cardType` from the code base, as everywhere it was in use it there is also a `Poll` object which holds the same information.
 - Added subviews to the `contentView` in `EmptyStateCell` rather than directly to that view.

Note: I did not address the `index` issue which you noted.

AAAstorga commented on Clicker/Models/ListDiffableModels/Poll.swift:
```diff
@@ -17,7 +17,8 @@ enum PollState {
 class Poll {
     
-    var id: Int?
+    let identifier = UUID().uuidString as NSString
```
why NSString?

jschluger replied:
Good question...removed.

AAAstorga commented on Clicker/SectionControllers/PollSectionController.swift:
```diff
@@ -27,12 +44,23 @@ class PollSectionController: ListSectionController {
     override func cellForItem(at index: Int) -> UICollectionViewCell {
-        let cell = collectionContext?.dequeueReusableCell(of: EmptyStateCell.self, for: self, at: index) as! EmptyStateCell
-        return cell
+        switch (userRole) {
```
get rid of parans here

AAAstorga commented on Clicker/SectionControllers/PollSectionController.swift:
```diff
@@ -27,12 +44,23 @@ class PollSectionController: ListSectionController {
+            socket.addDelegate(cell)
+            cell.configureWith(socket: socket, delegate: askedCardDelegate, poll: poll)
+            return cell
+        default:
```
should not be using default role. instead, if this is for the rest of the cases you can do `case .enum1, .enum2...`

jschluger replied:
How should I handle the `Switch must be exhaustive` error that appears if I only have cases for the two members of the enum since `userRole` is an optional? Default case that fails? Or a `guard let` to ensure `userRole` isn't `nil`?

AAAstorga commented on Clicker/ViewControllers/CardController+Extension.swift:
```diff
@@ -15,14 +15,17 @@ extension CardController: ListAdapterDataSource {
             }
         default:
-            return pollsDateArray.compactMap({ pollsDateModel -> PollDateModel? in
+            return pollsDateArray.enumerated().compactMap({ (i,pollsDateModel) -> PollDateModel? in
```
rename i to index and you can do `pollsDateArray.enumerated().compactMap { ... }` no need for extra set of `()`

AAAstorga commented on Clicker/ViewControllers/CardController+Extension.swift:
```diff
@@ -53,18 +72,50 @@ extension CardController: PollDateSectionControllerDelegate {
+        let newPoll = Poll(text: text, options: options, type: type, state: state)
+        appendPoll(poll: newPoll)
+        adapter.performUpdates(animated: true, completion: nil)
+        let lastIndexPath = IndexPath(item: 0, section: 0)//pollsDateArray[currentIndex].polls.count-1)
```
remove comment if it will not be used again, if it will then you can leave it

AAAstorga commented on Clicker/ViewControllers/CardController+Extension.swift:
```diff
@@ -53,18 +72,50 @@ extension CardController: PollDateSectionControllerDelegate {
+        let date = "today"
+        let newPollDate = PollsDateModel(date: date, polls: [poll])
+        
+        guard let _ = pollsDateArray else {
```
this is a good use case to do `if pollsDateArray == nil { ... }`

AAAstorga commented on Clicker/ViewControllers/CardController.swift:
```diff
@@ -277,6 +179,22 @@ class CardController: UIViewController {
+        case .vertical:
+            collectionViewLayout.scrollDirection = .vertical
+            setupVerticalNavBar()
+        default:
```
use `.horizontal` makes it more readable

AAAstorga commented on Clicker/Views/Cards/AskedCard.swift:
```diff
@@ -81,17 +89,8 @@ class AskedCard: UICollectionViewCell, CardDelegate, SocketDelegate {
-    
     func setupCard() {
-        switch (cardType) {
+        switch (poll.state) {
```
drop `()`

AAAstorga commented on Clicker/Views/Cards/EmptyStateCell.swift:
```diff
@@ -18,4 +42,67 @@ class EmptyStateCell: UICollectionViewCell {
+        setupViews()
+        setupConstraints()
+        // TODO: Add logic for setting up empty state or nonempty state
+        if (userRole == .admin && session.name == session.code) {
```
no need for parans here

AAAstorga commented on Clicker/Views/Cards/EmptyStateCell.swift:
```diff
@@ -18,4 +42,67 @@ class EmptyStateCell: UICollectionViewCell {
+        nameView.snp.makeConstraints { make in
+            make.width.equalTo(self.snp.width)
+            make.centerX.equalTo(self.snp.centerX)
+            make.top.equalTo(safeAreaLayoutGuide.snp.top)
```
!!!! good job using safeAreaLayoutGuide. Makes me happy

AAAstorga commented on Podfile:
```diff
@@ -1,5 +1,8 @@
 platform :ios, '9.0'
 
+# ignore all warnings from all pods
+inhibit_all_warnings!
```
I feel so much lol

jschluger replied:
Is this alright though? Not really much we can do about the warnings pods are raising, and they are really annoying (I find) when trying to look at my own errors.
